lib/screens/v2/candidate_generator/widgets/train_crowding.ex

```diff
@@ -63,36 +88,206 @@ defmodule Screens.V2.CandidateGenerator.Widgets.TrainCrowding do
       else
-        []
+        Enum.reverse(@ol_station_to_platform_map)
       end
+
+    previous_stop_index =
+      Enum.find_index(
+        ol_stop_sequence,
+        &(elem(&1, 0) == train_crowding_config.station_id)
+      ) - 1
+
+    {_, platform_id_tuple} = Enum.at(ol_stop_sequence, previous_stop_index)
+
+    relevant_platform_id = elem(platform_id_tuple, train_crowding_config.direction_id)
+
+    cached_prediction =
+      Agent.get(train_crowding_config.station_id, train_crowding_config.direction_id)
+
+    cond do
+      cached_prediction != nil ->
+        # Time-based heuristic
+        # We think the train is about to leave the previous station. Log the heuristic.
+        _ = check_time_based_heuristic(cached_prediction, common_params)
+
+        # Consecutive crowding class heuristic.
+        # Crowding class this fetch is the same as last. Log the heuristic.
+        _ =
+          check_consecutive_crowding_heuristic(
+            next_train_prediction,
+            relevant_platform_id,
+            cached_prediction,
+            common_params
+          )
+
+        []
+
+      # Start timer for dwell time but don't show the widget
+      Prediction.vehicle_status(next_train_prediction) == :stopped_at and
+          Prediction.stop_for_vehicle(next_train_prediction) == relevant_platform_id ->
+        previous_station_prediction_current_trip =
+          fetch_previous_station_prediction(
+            train_crowding_config,
+            relevant_platform_id,
+            next_train_prediction.trip.id,
+            common_params.fetch_predictions_fn
+          )
+
+        if is_nil(previous_station_prediction_current_trip) do
+          []
+        else
+          # Cache the departure_time from the previous station's prediction (minus 10 seconds for cushion).
+          # When now >= this time, show the widget.
+          Agent.put(
+            train_crowding_config.station_id,
+            train_crowding_config.direction_id,
+            previous_station_prediction_current_trip
+          )
+
+          []
+        end
+
+      true ->
+        []
+    end
+  end
+
+  defp fetch_previous_station_prediction(
+         train_crowding_config,
+         relevant_platform_id,
+         trip_id,
+         fetch_predictions_fn
+       ) do
+    params = %{
+      direction_id: train_crowding_config.direction_id,
+      route_ids: [train_crowding_config.route_id],
+      stop_ids: [relevant_platform_id],
+      trip_id: trip_id
+    }
+
+    {:ok, predictions} = fetch_predictions_fn.(params)
+    List.first(predictions)
+  end
+
+  defp check_time_based_heuristic(
+         cached_prediction,
+         common_params
+       ) do
+    show_widget_after_dt = DateTime.add(cached_prediction.departure_time, -10)
+
+    if DateTime.compare(
+         common_params.now,
+         show_widget_after_dt
+       ) in [
+         :eq,
+         :gt
+       ] do
+      log_crowding_info(
+        :dwell,
```

Comment:
Maybe adjust this to match the "rebrand" of time-based heuristic. Perhaps `:some_time_before_departure`

Reply:
Good catch. I went with `:time_based` because that's what I call the heuristic in code. Is that name clear or is it too general?

Reply:
I'm on the fence, but I suppose we can leave it for now (until we add another time based heuristic). But there is another comment referring to dwell time on line 284.